cloudevents/abstract/event.py

```diff
@@ -0,0 +1,97 @@
+import abc
+import typing
+from typing import TypeVar
+
+
+class CloudEvent(abc.ABC):
+    """
+    Used internally for multi-framework event type integration
+    """
+
+    @classmethod
+    @abc.abstractmethod
+    def create(
+        cls,
+        attributes: typing.Dict[str, typing.Any],
+        data: typing.Optional[typing.Any],
+    ) -> "CloudEvent":
+        """
+        Factory function.
+        We SHOULD NOT use the __init__ function as the factory because
+        we MUST NOT assume how the __init__ function looks across different python
+        frameworks
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def _attributes_read_model(self) -> typing.Dict[str, typing.Any]:
+        """
+        :return: attributes of this event
+
+        you MUST NOT mutate this dict
+        implementation MAY assume the dic will not be mutated
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def _data_read_model(self) -> typing.Optional[typing.Any]:
+        """
+        :return: data value of the  event
+        you MUST NOT mutate this value
+        implementation MAY assume the value will not be mutated
+        """
+        pass
+
+    def __setitem__(self, key: str, value: typing.Any) -> None:
+        raise NotImplementedError()
```

Comment:
Can't we use `self._data_read_model()[key]=value`?

I don't really get why we want to have a `read` model.

Reply:
It MAY not be possible to do so. 
For example - if my cloudevent is a database row, or a cached value.
I can read the value, but setting it MAY not be possible in the same way.

updated the docs so it will be more clear

Reply:
Again, it is up to us to define the contract. I'd rather have the default implementation here and rely on whatever we present in the contract.
It's kinda the implementation problem of how to present the data.
While we're kinda obliged to follow the existing API till we move to v2, I'd rather keep it simple and leave it to the implementation to follow.

I'd probably go with a fully immutable class in v2 and then define some mutation helpers, but that's a different story.

Reply:
It's pretty simple to read one way and write another way
its CQRS.
if you wish to change it, i can

Reply:
I don't think this is cqrs. Haven't ever heard of using it in the context of a single entity.

We may not provide a mutation API at all at this level.

Reply:
I think that you are right. i will remove the del and set from this level